csharp/src/Microsoft.ML.OnnxRuntime/OrtValue.shared.cs

```diff
@@ -847,56 +884,340 @@ public static OrtValue CreateFromStringTensor(Tensor<string> tensor)
         /// </summary>
-        /// <param name="ortValues">a collection of OrtValues</param>
+        /// <param name="ortValues">a collection of OrtValues. On success the ortValues contained in the list
+        /// are taken ownership of and the list is cleared.</param>
         /// <returns>A disposable instance of OrtValues</returns>
         /// <exception cref="ArgumentNullException"></exception>
-        public static OrtValue CreateSequence(IReadOnlyCollection<OrtValue> ortValues)
+        public static OrtValue CreateSequence(ICollection<OrtValue> ortValues)
         {
             if (ortValues is null)
             {
                 throw new ArgumentNullException(nameof(ortValues));
             }
 
-            var handles = new IntPtr[ortValues.Count];
-            for (int i = 0; i < ortValues.Count; i++)
+            if (ortValues.IsReadOnly)
+            {
+                throw new ArgumentException("ortValues argument can not be a readonly collection");
+            }
+
+            var compositeMembers = new DisposableList<OrtValue>(ortValues);
+            try
             {
-                handles[i] = ortValues.ElementAt(i).Handle;
+                var result = CreateSequence(ref compositeMembers);
+                Debug.Assert(compositeMembers is null, "Must be null on success");
+                ortValues.Clear();
+                return result;
+            }
+            catch (Exception)
+            {
+                // The caller is responsible for disposing the ortValues
+                compositeMembers?.Clear();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Creates a sequence from the values in compositeMembers
+        /// The argument is taken possession of and is nullified on successful return.
+        /// </summary>
+        /// <param name="compositeMembers">sequence ortValues</param>
+        /// <returns>OrtValue instance representing a Sequence</returns>
+        internal static OrtValue CreateSequence(ref DisposableList<OrtValue> compositeMembers)
+        {
+            var handles = new IntPtr[compositeMembers.Count];
+            for (int i = 0; i < compositeMembers.Count; i++)
+            {
+                handles[i] = compositeMembers[i].Handle;
             }
 
             NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateValue(handles,
-                (UIntPtr)ortValues.Count, (IntPtr)OnnxValueType.ONNX_TYPE_SEQUENCE,
-                out IntPtr sequenceHandle));
-            return new OrtValue(sequenceHandle, OnnxValueType.ONNX_TYPE_SEQUENCE);
+                    (UIntPtr)handles.Length, (IntPtr)OnnxValueType.ONNX_TYPE_SEQUENCE,
+                    out IntPtr sequenceHandle));
+
+            return new OrtValue(sequenceHandle, OnnxValueType.ONNX_TYPE_SEQUENCE, ref compositeMembers);
         }
 
+        /// <summary>
+        /// A delegate type that is expected to process each OrtValue in a sequence.
```

Comment:
>A delegate type that [](http://example.com/codeflow?start=12&length=20)

Would it be beneficial for the visitor to return a boolean to stop visitation?

Reply:
What's the expected usage and how often would you want to exit early?

If you're iterating results I assume you need to iterate all of them.

Reply:
I also think so, but some examples on .ai site get either the last or the first. It can either be contrived, or people can always use GetValue() directly.